src/v/cluster/producer_state.cc

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "producer_state.h"
+
+#include "cluster/logger.h"
+#include "cluster/producer_state_manager.h"
+#include "vassert.h"
+
+namespace cluster {
+
+result_promise_t::future_type request::result() const {
+    return _result.get_shared_future();
+}
+
+std::optional<request_ptr> requests::last_request() const {
+    if (!_inflight_requests.empty()) {
+        return _inflight_requests.back();
+    } else if (!_finished_requests.empty()) {
+        return _finished_requests.back();
+    }
+    return std::nullopt;
+}
+
+bool requests::is_valid_sequence(seq_t incoming) const {
+    auto last_req = last_request();
+    return
+      // this is the first request with seq=0
+      (!last_req && incoming == 0)
+      // incoming request forms a sequence with last_request
+      || (last_req && last_req.value()->_last_sequence + 1 == incoming)
+      // sequence numbers got rolled over because they hit int32 max limit.
+      || (last_req && last_req.value()->_last_sequence == std::numeric_limits<seq_t>::max() && incoming == 0);
+}
+
+result<request_ptr> requests::try_emplace(
+  seq_t first, seq_t last, model::term_id current, bool reset) {
```

Comment:
it's better to be light on aliases because they get in the way of searching things in the code, in this case the last parameter has two names `reset` and `reset_sequences` making `git grep` useless. I've already ran into this problem trying to find how the flag is used in this PR with `Ctrl-F`